internal/rbd/snap_diff.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2025 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	rbderrors "github.com/ceph/ceph-csi/internal/rbd/errors"
+	"github.com/ceph/ceph-csi/internal/rbd/types"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc/codes"
+)
+
+// StreamMetadata streams the block metadata for a snapshot.
+// If baseRBDSnap is provided, it calculates the delta between the
+// current snapshot and the base snapshot.
+func (rbdSnap *rbdSnapshot) StreamMetadata(
+	ctx context.Context,
+	sendResponse types.SMSSendResponse,
+	startingOffset int64,
+	maxResults int32,
+	baseSnap types.Snapshot,
+) error {
+	var fromSnapID uint64
+	image, err := rbdSnap.open()
+	if err != nil {
+		return fmt.Errorf("failed to open image %q: %w", rbdSnap, err)
+	}
+	defer func() {
+		cErr := image.Close()
+		if cErr != nil {
+			log.WarningLog(ctx, "resource leak, failed to close image: %v", cErr)
```

Comment:
log it as error?

Reply:
I think of errors as fatal, a warning for a resource leak (unlikely to happen) is more suitable.